provider/vsphere/internal/vsphereclient/client.go

```diff
@@ -0,0 +1,327 @@
+	}
+	return errors.Annotate(lastError, "failed to remove instances")
+}
+
+// VirtualMachines return list of all VMs in the system matching the given path.
+func (c *Client) VirtualMachines(ctx context.Context, path string) ([]*mo.VirtualMachine, error) {
+	finder, _, err := c.finder(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	items, err := finder.VirtualMachineList(ctx, path)
+	if err != nil {
+		if _, ok := err.(*find.NotFoundError); ok {
+			return nil, nil
+		}
+		return nil, errors.Annotate(err, "listing VMs")
+	}
+
+	vms := make([]*mo.VirtualMachine, 0, len(items))
+	for _, item := range items {
+		var vm mo.VirtualMachine
+		err := c.client.RetrieveOne(ctx, item.Reference(), nil, &vm)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		vms = append(vms, &vm)
+	}
+	return vms, nil
+}
+
+// ComputeResources retuns list of all root compute resources in the system.
+func (c *Client) ComputeResources(ctx context.Context) ([]*mo.ComputeResource, error) {
+	_, datacenter, err := c.finder(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	folders, err := datacenter.Folders(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	root := list.Element{Object: folders.HostFolder}
+	es, err := c.recurser().Recurse(ctx, root, []string{"*"})
+	if err != nil {
+		return nil, err
+	}
+
+	var cprs []*mo.ComputeResource
+	for _, e := range es {
+		switch o := e.Object.(type) {
+		case mo.ClusterComputeResource:
+			cprs = append(cprs, &o.ComputeResource)
+		case mo.ComputeResource:
+			cprs = append(cprs, &o)
+		}
+	}
+	return cprs, nil
+}
+
+// EnsureVMFolder creates the a VM folder with the given path if it doesn't
+// already exist.
+func (c *Client) EnsureVMFolder(ctx context.Context, folderPath string) error {
+	finder, datacenter, err := c.finder(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	folders, err := datacenter.Folders(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	createFolder := func(parent *object.Folder, name string) (*object.Folder, error) {
+		folder, err := parent.CreateFolder(ctx, name)
+		if err != nil && soap.IsSoapFault(err) {
+			switch soap.ToSoapFault(err).VimFault().(type) {
+			case types.DuplicateName:
+				return finder.Folder(ctx, parent.InventoryPath+"/"+name)
+			}
+		}
+		return folder, err
+	}
+
+	parentFolder := folders.VmFolder
+	for _, name := range strings.Split(folderPath, "/") {
+		folder, err := createFolder(parentFolder, name)
+		if err != nil {
+			return errors.Annotatef(
+				err, "creating folder %q in %q",
+				name, parentFolder.InventoryPath,
+			)
+		}
+		parentFolder = folder
+	}
+	return nil
+}
+
+// DestroyVMFolder destroys a folder rooted at the datacenter's base VM folder.
+func (c *Client) DestroyVMFolder(ctx context.Context, folderPath string) error {
+	finder, datacenter, err := c.finder(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	folders, err := datacenter.Folders(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
```

Comment:
This seems to be duplicated in couple of places... Maybe re-factor into one call?